### ISSUE TITLE
Support readOnly and writeOnly openapi properties

### ISSUE THREAD

**Issue by m-vellinga** [koxudaxi/datamodel-code-generator#727]
**Is your feature request related to a problem? Please describe.**
OpenAPI allows you to set a `readOnly` & `writeOnly` for fields to be able to reuse schemas for get and post for instance. See [docs](https://swagger.io/docs/specification/data-models/data-types/#readonly-writeonly) for more information. Would be nice to also support this, makes things a lot more "dry" on the spec side of things.

**Describe the solution you'd like**
Since Pydantic does *not* support read/write only fields a possible solution could be to generate 2 versions of the same schema one with and one without the read/write only field.

**Describe alternatives you've considered**
Alternative would be to define 2 different schemas in the openapi spec instead

**Comment by koxudaxi** [comment]
@m-vellinga 
Thank you for suggesting the feature.
How do you control `read/write` in the pydantic world?
Could you please tell the detail clearly?

**Comment by karolzlot** [comment]
What is possible to have in pydantic is `allow_mutation` Field constraint (default is `True`).
And this is my use case, I also need to be able to set `allow_mutation=False` for some fields in generated pydantic models.

Now how we can define it in JSON Schema:
1. `readOnly`

   - Pros:
      - It is already in JSON Schema specification
   - Cons: 
       - Misleading name, at least according to samuelcolvin. He didn't agree to generate `readOnly` when generating JSON Schema from pydantic models, you can see in those two comments:
https://github.com/samuelcolvin/pydantic/pull/2196#discussion_r550794426
https://github.com/samuelcolvin/pydantic/pull/2196#discussion_r550794616
This is reverted commit: 
https://github.com/samuelcolvin/pydantic/pull/2196/commits/521d254e400f588a2c48bdea7c38bc092dfd32c3#diff-c7021feccdf7c6e5a107ebfb58d3d553f2e30ab82be8acc88738c031f032b1ccL281

2. `immutable` 


   - Pros: 
      - is more clear compared to `readOnly`, at least according to samuelcolvin
   - Cons: 
      - not yet in JSON Schema specification
   - proposed here: https://github.com/json-schema-org/json-schema-vocabularies/issues/31
   - also proposed for OpenAPI-Specification: https://github.com/OAI/OpenAPI-Specification/issues/2720

3. `createOnly`
   - suggested here: https://github.com/OAI/OpenAPI-Specification/issues/2720#issuecomment-925184694

However please also read all comments here: https://github.com/OAI/OpenAPI-Specification/issues/2720
It seems that choosing right name for it is really not easy...

Overall I'm not sure about correct decision here. Samuelcolvin clearly would chose `immutable`, but I'm not sure if he is right. I think that names in JSON Schema and in pydantic doesn't need to match, because they doesn't represent the same things.
But I'm not sure, I need to probably think about it more...

----

For now we can just choose any name of those and change it later when this feature will be added to JSON Schema specification.

**Comment by karolzlot** [comment]
Worth reading:
chapter in JSON Schema specification about "readOnly" and "writeOnly":
https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.9.4

**Comment by karolzlot** [comment]
According to what I read in this specification, we should have not only "readOnly" and "writeOnly", but also:


Ok, after writing [this comment with a table](https://github.com/OAI/OpenAPI-Specification/issues/2720#issuecomment-1121925715) I came to conclusion that:

samuelcolvin is right and `readOnly` isn't a good choice.

From pydantic's point of view both `immutable` and `createOnly` are ok and should be coverted to pydantic's `allow_mutation=False`, but I think if we need to choose one then `immutable` is more intuitive.

**Comment by karolzlot** [comment]
Ok, after thinking more about this I again have doubts....
Maybe `readOnly` is the best choice after all?

**Comment by rafalkrupinski** [comment]
I think read/write-only properties are more about if they're allowed in object.
Read-only property should be only allowed in a response model, so there should be validation for the request model that would forbid it
Also, the property should be optional even if it's required, so a read-only property can be None in response object and vice versa

**Comment by mattholden** [comment]
Plus-one on this request. Currently every endpoint in my API has to have three separate schemas, for get, post, and patch. This makes development of our product far more complicated than it has to be, for those fields being missing. The complexity of our API is reaching a point where DMCG is choking on it for versions >= 18, meaning we're blocked on Pydantic 2 and FastAPI 100. 

I don't even need them to be immutable in pydantic, I just need the system to know that if a property is `readOnly` it is only allowed in responses, and if it's `writeOnly` it's only allowed in request bodies.

**Comment by heath-freenome** [comment]
I would also love them to be output in the `msgspec` version as well. Even if they are only put into the `extras` dict

**Comment by CharString** [comment]
`msgspec` user here. I came, because I discovered that `readOnly` made attributes `Optional`.
Not fun, because it makes the type checker force us to handle `None` cases that don't exist, and accept incomplete responses from the API, and allows us to send `readOnly` attributes to the API, possibly triggering 4xx errors.

> I think read/write-only properties are more about if they're allowed in object. Read-only property should be only allowed in a response model, so there should be validation for the request model that would forbid it

Which means not having the attribute on the class at all. This is not `msgspec` specific, but type system specific. The same holds for Pydantic Models, if you want your type checker to be of any help and not be a nuisance. `msgspec` does distinguish between `None` and `UNSET`, but there's no such thing as a `FORBIDDEN` value and type. 

> Also, the property should be optional even if it's required, so a read-only property can be None in response object and vice versa

No, if it is required it should be present. Whether it's nullable is orthogonal to this aspect of `readOnly`. The read only version of the type is a similar, but different type than the write version. You could expres this with inheritance e.g.:

```python
class WritableMyType(Struct):
    name: str

class MyType(WritableMyType, frozen=True):
    id: uuid
```

The `frozen=True` part make the whole Struct pseudo-immutable, much like `allow_mutation=False` mentioned, but on the whole Struct. Just like dataclasses.

**Comment by rafalkrupinski** [comment]
Perhaps I didn't express myself clearly.
in JSON, a property can be absent from an object even if it's required, when it's readOnly or writeOnly. In python, it's a matter of translation of type systems: when my schema has 1000 of objects, I might not want 3000 python classes (base + read + write) but rather use `Optional[]` fields, at the cost of not being strictly correct.

**Comment by rafalkrupinski** [comment]
I think that `MAY` is there because whether a property is required depends on `required`, `readOnly`/`writeOnly` and the direction of the message (request/response). If the conditions for being required are met, that property must be present. Don't ask me to quote specs :)

**Comment by CharString** [comment]
Why define the types at all then? Just tab-completion in your IDE?

**Comment by St0rmdrag0n** [comment]
Yeah, currently, we have to define three versions of the nearly-same model - response, create request, and update request. It just makes things a lot messier than it needs to be and not very DRY. We'd love to be able to say, "This is a Foo object, which has 40 fields. Of those, 10 are generated on our side and you'll never send them, and 5 of them you can pass on create but not on update".

**Comment by St0rmdrag0n** [comment]
This is especially painful for us because the same OpenAPI spec we use to generate our models with DMCG is also used to generate our docs, and we also WANT to use it to generate SDK libraries for our end-users. Having three separate models - at least - for every resource really muddies that other tooling up, too.

**Comment by Diegovsky** [comment]
Let's follow the least surprising path and do like [openapi-ts](https://github.com/hey-api/openapi-ts) does, which is to generate a separate "model" for each case of `create`, `get` and `update`. It's a matter of prefixing the "original/base" schema with its intent.

I would be down to a PR to implement this in pydantic v3, which is what I use. What do you think @koxudaxi ?

**Comment by rafalkrupinski** [comment]
I don't mean to be picky, but lets avoid confusion.
There's no such thing as "create", "get" or "update" in OpenAPI. There are just REST operations, and they may have a schemas with `readOnly` and `writeOnly` properties, and you may need to generate separate `Request` and `Response` models if a schema (or any property schema) defines `read-` or `writeOnly` property.

Whether or not `update` operation defines a separate schema, is completely another matter, but OpenAPI doesn't provide a good way to derive one from other schemas used by `get` or `create` operations.

**Comment by Diegovsky** [comment]
Yeah, being pedantic is okay, but you understood what I meant. It usually is the case that a CRUD endpoint defines 3 models each, and @St0rmdrag0n did mention it is their use case.

In any case, I think generating separate types for `Read<Schema>`, `Write<Schema>` and just `<Schema>` probably fits 99% of cases. What do you think?

**Comment by rafalkrupinski** [comment]
I think generating between one and three models, just like you wrote, depending on the presence of `readOnly` and `writeOnly`, will cover 100% of OpenAPI 3.x.

BTW, I strongly suggest using `Request`/`Response` naming pattern, as `read/write` is ambiguous (what's written out by the client is read by server, and vice-versa, so both models can be interpreted as `read` and `write` unless you know from the OAS which is which).

I'm not sure if @St0rmdrag0n's use case is the same.
For a single resource model, an object schema field in `create` operation may be required and non-nullable, while in `update` it may be optional and nullable. It's not the same as "this field is absent in requests" as in case of `readOnly`. One may be tempted to use a complex schema with `allOf`, but that's not a good idea, and I haven't seen it done properly.
And then there's the python side, where generating a super- and sub-class isn't really a good representation of such complex schemas.
